board.cpp

```diff
@@ -40,41 +43,111 @@ QVariant Board::data(const QModelIndex &index, int role) const
+        m_state = BoardState::Select;
+    }
+
+    switch (m_state) {
+    case BoardState::Select:
+        m_firstItem = {inx / m_dimension, inx % m_dimension};
+        m_data[m_firstItem.x()][m_firstItem.y()].setSelected(true);
+        emit dataChanged(this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                         this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                         {Qt::UserRole + 2});
+        m_state = BoardState::Swap;
+        break;
+    case BoardState::Swap:
+        m_secondItem = {inx / m_dimension, inx % m_dimension};
+        if (!isMovable(m_firstItem.x() * m_dimension + m_firstItem.y(),
+                   m_secondItem.x() * m_dimension + m_secondItem.y()))
+        {
+            m_data[m_firstItem.x()][m_firstItem.y()].setSelected(false);
+            emit dataChanged(this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                             this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                             {Qt::UserRole + 2});
+            m_firstItem = m_secondItem;
+            m_data[m_firstItem.x()][m_firstItem.y()].setSelected(true);
+            emit dataChanged(this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                             this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                             {Qt::UserRole + 2});
+        } else {
+            m_data[m_firstItem.x()][m_firstItem.y()].setSelected(false);
+            emit dataChanged(this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                             this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                             {Qt::UserRole + 2});
+            move();
+            m_state = BoardState::Pop;
+        }
+        break;
+    case BoardState::Pop:
+        if (!pop())
+        {
+            move();
+            m_state = BoardState::None;
+        } else {
+            m_state = BoardState::Shift;
+            doSomething();
+        }
+        break;
+    case BoardState::Shift:
+        m_state = BoardState::Fill;
+        if (!shift())
+        {
+            doSomething();
+        }
+        break;
+    case BoardState::Fill:
+        fill();
+        m_state = BoardState::None;
+        break;
+    default:
+        m_firstItem = cInvalidPoint;
+        break;
     }
 }
 
+bool Board::move()
+{
+    Tile::swapPosition(m_data[m_firstItem.x()][m_firstItem.y()], m_data[m_secondItem.x()][m_secondItem.y()]);
+    emit dataChanged(this->index(m_firstItem.x() * m_dimension + m_firstItem.y()),
+                     this->index(m_firstItem.x() * m_dimension + m_firstItem.y()));
+    emit dataChanged(this->index(m_secondItem.x() * m_dimension + m_secondItem.y()),
+                     this->index(m_secondItem.x() * m_dimension + m_secondItem.y()));
+    std::swap(m_data[m_firstItem.x()][m_firstItem.y()], m_data[m_secondItem.x()][m_secondItem.y()]);
+    return true;
```

Comment:
is there a case when this function returns false?
As I can see we don't need return value anymore